File: Data-Driven Polynomial Optimization/Julia Scripts/MG_LyapFn.jl
# -------------------------------------------------------------------------
# Identifying Lyapunov Functions with Semidefinite Programming
#
# The goal of this script is to use semidefinite programming to identify a
# Lyapunov function for the planar Moore-Greitzer system
#      x1' = -x2 - 1.5x1^2 - 0.5x1^3,
#      x2' = 3x1 - x2.
# The script expands the unknown Lyapunov function as a polynomial up to a
# given degree and tunes the coefficients in order to satisfy both of the
# constraints:
#      V - x1^2 - x2^2 is SOS
#      -LV - x1^2 - x1^2 is SOS
# where LV is the Lie derivative associated to the Moore-Greitzer system.
#
# To run this notebook one requires paths to the freely available software
# package MOSEK, which can be downloaded at:
#           https://www.mosek.com/downloads/
#
# This script accompanies Section 4.2 of Data-Driven Methods for
# Dynamic Systems.
#
# This script was adapted from the MATLAB version by Jason J. Bramburger
# 
# Author: Mohit Sahu
# -------------------------------------------------------------------------


using DynamicPolynomials, SumOfSquares
using MosekTools
using CSDP

@polyvar x[1:2]

f = [-x[2] - 1.5*x[1]^2 - 0.5*x[1]^3;
      3*x[1] - x[2]]

# solver = optimizer_with_attributes(CSDP.Optimizer, MOI.Silent() => true)
model = Model(Mosek.Optimizer)
V_basis = monomials(x, 0:6)  # monomials for polynomial V
@variable(model, c[1:length(V_basis)])

# As julia doesn't handle L1 norm, so t here is a dummy variable to solve this problem
# See Tutorials/Linear programs/Tips and Tricks of Jump.jl documentation to know more about this
@variable(model, t)

V = c'*V_basis

using LinearAlgebra
Dv = differentiate(V, x)
Lie = dot(Dv, f)

# Add SOS constraints
@constraints(model, begin
    V .- sum(x.^2) in SOSCone()
    -Lie - sum(x.^2) in SOSCone()
    [t; c] in MOI.NormOneCone(1+length(c))   # L1 Norm
end)

@objective(model, Min, t)
optimize!(model)

# Check if a solution was found
if termination_status(model) == MOI.OPTIMAL
    println("Lyapunov function identified!")
    V_sol = value(V)
    threshold = 1e-1
    filtered_coeffs = [abs(c) > threshold ? c : zero(c) for c in coefficients(V_sol)]
    println("Filtered V(x1, x2) = ", sum(filtered_coeffs .* monomials(V_sol)))
else
    println("Sorry, unable to identify a Lyapunov function")
end

# Plot the Lyapunov function
using Plots
plotlyjs()
N = 1000;
x1_range = LinRange(-5, 5, N)

x = x1_range' .* ones(length(x1_range))
y = ones(length(x1_range))' .* x1_range
V_discovered = value.(c)'*V_basis;

plot(x, y, V_discovered.(x,y), st=:surface)
xlabel!("x1")
ylabel!("x2")
zlabel!("V(x1, x2)")

File: Data-Driven Polynomial Optimization/MG_LyapFn.m
% -------------------------------------------------------------------------
% Identifying Lyapunov Functions with Semidefinite Programming
%
% The goal of this script is to use semidefinite programming to identify a
% Lyapunov function for the planar Moore-Greitzer system
%      x1' = -x2 - 1.5x1^2 - 0.5x1^3, 
%      x2' = 3x1 - x2.
% The script expands the unknown Lyapunov function as a polynomial up to a
% given degree and tunes the coefficients in order to satisfy both of the
% constraints:
%      V - x1^2 - x2^2 is SOS
%      -LV - x1^2 - x1^2 is SOS
% where LV is the Lie derivative associated to the Moore-Greitzer system.
%
% To run this notebook one requires paths to the freely available software
% packages YALMIP and MOSEK. YALMIP can be downloaded at:
%           https://yalmip.github.io/download/
% and MOSEK can be downloaded at:
%           https://www.mosek.com/downloads/
%
% This script accompanies Section 4.2 of Data-Driven Methods for
% Dynamic Systems. 
%
% Author: Jason J. Bramburger
% -------------------------------------------------------------------------

% Clean workspace
clear all; 
close all;
yalmip clear
clc

%% Discovering polynomial Lyapunov function

% Degree of Lyapunov function
degV = 6;

% Symbolic variables
x = sdpvar(2,1);

% Moore-Greitzer right-hand-side
f = [-x(2) - 1.5*x(1)^2 - 0.5*x(1)^3; 3*x(1) - x(2)];

% Lyapunov function, gradient, and Lie derivative
[v,vc,vm] = polynomial(x,degV,1); 
Dv = jacobian(v,x);        
Lie = dot(Dv,f);

% Find Lyapunov function
cons = [sos(v - dot(x,x)); sos(-Lie - dot(x,x))]; % SOS constraints to be satisfied 
obj = sum(abs(vc)); % objective function: minimize l^1 norm of coefficients 
sol = solvesos(cons,obj,[],vc);

%% Print Results

% Clean window
clc

% Print if solution was solved
if sol.problem == 0 % <-- solution is solved successfully
    
    % Remove out small values from coefficients
    coeffs = clean(value(vc),1e-5);
    
    % Display Lyapunov function
    fprintf('Lyapunov function identified! \n')
    fprintf('\n')
    fprintf('V(x(1),x(2)) = ') 
    sdisplay(coeffs.'*vm)
else
    % If no Lyapunov function, tell the user
    fprintf('Sorry, unable to identify a Lyapunov function \n')
end

%% Plot Lyapunov function

% Spatial coordinates
N = 1000;
x2 = linspace(-5,5,N);
y2 = linspace(-5,5,N);
[X,Y] = meshgrid(x2,y2);

% Build Lyapunov function
v_num = dot(value(vc),vm);
[pow, coeff] = getexponentbase(v_num,x);
lyap = zeros(size(X));
for i = 1:numel(lyap)
    lyap(i) = lyap(i) + coeff*( X(i).^pow(:,1) .* Y(i).^pow(:,2) );
end

% Surface plot
surf(X,Y,lyap)
shading interp
xlabel('$x_1$','Interpreter','Latex')
ylabel('$x_2$','Interpreter','Latex')
zlabel('$V(x_1,x_2)$','Interpreter','Latex')
set(gca,'FontSize',16,'Xlim',[-5 5],'Ylim',[-5 5])


